### DISCUSSION TITLE
Add on with timed off configuration to lights

### DISCUSSION THREAD

**Discussion by Kane610** [home-assistant/architecture#615] [Entity Models]
I would like to propose to add a new configurable parameter to lights (most likely it would fit just the same in the switch platform); an on with timed off.

This would extend the current schema with a positive float representing seconds which the light should be turned on before being turned off again. This would be more or less a copy/paste of the Transition parameter in the current implementation.

And just as with transition it would be up to each integration to manage how to send the command.

```
SUPPORT_ON_WITH_TIMED_OFF = 64
VALID_ON_WITH_TIMED_OFF = vol.All(vol.Coerce(float), vol.Clamp(min=0, max=6553))
LIGHT_TURN_ON_SCHEMA = {..., ATTR_ON_WITH_TIMED_OFF: VALID_ON_WITH_TIMED_OFF, ...}

filter_turn_on_params / filter_turn_off_params
if not supported_features & SUPPORT_ON_WITH_TIMED_OFF:
    params.pop(ATTR_ON_WITH_TIMED_OFF, None)
```

Integrations that could utilise this
- deCONZ
- ZHA
- ZWave
- Nedis
- Shelly
- Tasmota
- ESPHOME
- Xiaomi Miio (lists something like Delayed turn off)
- Yeelight (lists a service YEELIGHT.SET_AUTO_DELAY_OFF_SCENE)

**Comment by shbatm** [comment]
I would like to see this as well, but extended to more than just the light domain. I have several switches I've written custom automations for turning off after a certain time, but it requires additional entities to be created for each device (timer, input_number) a script, and 2 automations to handle the cancelled or finished timer. It would be nice if this could be handled completely in the background by the Core.

**Reply by Kane610** [reply]
Absolutely, it might even be relevant for other entity types building on top of the toggle entity.

**Comment by rytilahti** [comment]
Wouldn't it be easier to support this through a generic blueprint / custom "proxy" integration for "turn on for <timeperiod>" than adding a new feature to integrations (and different platforms), even when plenty of devices support some sort of delayed off feature?

**Reply by Kane610** [reply]
I don't understand how this would be easier. You can combine the on_time attribute with most other light attributes, so it would need to be a part of the light turn on service.

**Reply by frenck** [reply]
I agree, with @rytilahti. I guess this is a feature that becomes problematic when mixing and matching different types of lights (ones that support it, and ones that don't), causing the behavior to becomes really unexpected in those cases.

Next, the cases arise that should be able to reset the time during the time is running (e.g., when a motion sensors re-triggers).

In the end, this already can be done using automation and script, in more universal and advanced ways that probably will never be possible with the on_time proposal.

**Reply by thecode** [reply]
I think that a main point is being missed here, many devices support hardware timer, I know the discussion is focused towards lights, but as already suggested should be looked in more general approach.

Two integrations I am maintaining - Shelly & Switcher both supports hardware timer for auto off, in Shelly all devices support auto off timer so can be used for switch entities. In Switcher it is very important to use the hardware timer since the device is used to control a water heater and you don't want the heater to stay on too long in case the device lose WIFI or server is down. I am also using several Tasmota devices for irrigation valves control, again with a hardware timer (same here, an automation fault will leave water running forever)

A blueprint or automation is not the ideal solution to this if the device has hardware support, of course the integration can support custom service for it (what we do in Switcher) but would be better if it is a part of the `turn_on` command for all toggle based entities.

I agree there is a need to reset the timer or disable a running timer, for both Switcher and Shelly calling the timer again with a new value will replace the value, calling it with 0 or turning off the switch will also disable the timer. It might not be the same for all devices, but we can define & document the required behavior and let integrations deal with what is needed to support a timer. E.g. if a device has a dedicated API for resetting the timer (not by directly sending a new value) the integration can detect a change in the timer setting and call the relevant API to set a new value.

**Reply by Kane610** [reply]
Normal case is that the timer is reset when called with a new time and removed when called without time. This could be clarified in the dev docs that the behavior is intended to be kept similar between integrations.

**Comment by Kane610** [comment]
So ok to go? This is really similar to the transition functionality which also is quite common in integrations.

**Reply by MartinHjelmare** [reply]
I think we need a couple of more positive members first since some concerns have been raised.

Also, I'm missing an explicit description of the suggested implementation in the top post. A list of all known integrations that can support this is good to add, so we can update that until the suggestion is approved or rejected.

**Comment by rytilahti** [comment]
So, I've been thinking on this, and I'm wondering if it would make more sense to add a separate "standard action" (or w/e naming would fit the bill) for such common features, instead of overloading the `turn_on`/`turn_off` even more?

In the end, delayed turn off is just a one-shot scheduled action. As user I don't really care if it's done on software or on hardware after all. I just want to schedule something to happen in 30 mins or at 8 o'clock, how some device implements this doesn't really matter to me that much.

To put it in other words, would it rather make more sense to implement a higher-level facility to deal with scheduling? Maybe a generic `schedule` component that offers this feature for all controllable entities, while allowing integrations to provide `schedule` platform to handle hardware-based scheduling?

**Comment by Bre77** [comment]
Just another data point, I am the code-owner of **Advantage Air**, and that product has the hardware on/off timer for a climate entity, so the concept of a `schedule` platform would work in my case assuming its entity/platform independant. Right now I implement this feature using sensors (showing the current countdown) and custom services.

I generally think adding platform level support to light for on/off timer is good, as long as its generic to the platform, and integrations can overload the built in function with an integration specific one using hardware scheduling where thats possible.

**Reply by rytilahti** [reply]
So, in my mind the `schedule` (or `timer`, to keep it more tightly scoped?) should be a core integration that offers the scheduling no matter the integration:
* If an integration implements the necessary interfaces (as a platform) to control hardware timers and sets a flag them being supported, great!
* As a fallback for integrations that do not support this feature, the integration will take care of it by providing a software implementation.

Note that what I'm writing here is likely just a very naive approach, and this is merely how I personally think how this should be implemented to provide a common interface for automations like "turn off after 30 minutes" and "turn on at 5am".

In my opinion the `turn_{off,on}` should not be overloaded to cater this use case, for the simple reason that the transition is directly bound to the action being performed where *scheduling* an action is much different from it.

**Reply by Kane610** [reply]
@rytilahti it is also important part of the command to allow setting other aspects of the device, brightness, color, etc.

The Positive part of this suggestion is to normalise the controls, where as if my suggestion would be accepted would implement the same mechanism in multiple platforms (light, switch, others?)
